plugin/backend_linux.go

```diff
@@ -61,59 +63,137 @@ func (pm *Manager) Inspect(name string) (tp types.Plugin, err error) {
+	var privileges types.PluginPrivileges
+	if c.Network.Type != "null" && c.Network.Type != "bridge" {
+		privileges = append(privileges, types.PluginPrivilege{
+			Name:        "network",
+			Description: "permissions to access a network",
+			Value:       []string{c.Network.Type},
+		})
+	}
+	for _, mount := range c.Mounts {
+		if mount.Source != nil {
+			privileges = append(privileges, types.PluginPrivilege{
+				Name:        "mount",
+				Description: "host path to mount",
+				Value:       []string{*mount.Source},
+			})
+		}
+	}
+	for _, device := range c.Linux.Devices {
+		if device.Path != nil {
+			privileges = append(privileges, types.PluginPrivilege{
+				Name:        "device",
+				Description: "host device to access",
+				Value:       []string{*device.Path},
+			})
+		}
+	}
+	if c.Linux.DeviceCreation {
+		privileges = append(privileges, types.PluginPrivilege{
+			Name:        "device-creation",
+			Description: "allow creating devices inside plugin",
+			Value:       []string{"true"},
+		})
+	}
+	if len(c.Linux.Capabilities) > 0 {
+		privileges = append(privileges, types.PluginPrivilege{
+			Name:        "capabilities",
+			Description: "list of additional capabilities required",
+			Value:       c.Linux.Capabilities,
+		})
+	}
+
+	return privileges, nil
 }
 
-// Pull pulls a plugin and computes the privileges required to install it.
-func (pm *Manager) Pull(name string, metaHeader http.Header, authConfig *types.AuthConfig) (types.PluginPrivileges, error) {
-	ref, err := distribution.GetRef(name)
+// Privileges pulls a plugin config and computes the privileges required to install it.
+func (pm *Manager) Privileges(name string, metaHeader http.Header, authConfig *types.AuthConfig) (types.PluginPrivileges, error) {
+	_, pd, err := pm.pull(name, metaHeader, authConfig)
 	if err != nil {
-		logrus.Debugf("error in distribution.GetRef: %v", err)
 		return nil, err
 	}
-	name = ref.String()
+	return computePrivileges(pd)
+}
 
-	if p, _ := pm.pluginStore.GetByName(name); p != nil {
-		logrus.Debug("plugin already exists")
-		return nil, fmt.Errorf("%s exists", name)
+// Pull pulls a plugin, check if the correct privileges are provided and install the plugin.
+func (pm *Manager) Pull(name string, metaHeader http.Header, authConfig *types.AuthConfig, privileges types.PluginPrivileges) error {
+	ref, pd, err := pm.pull(name, metaHeader, authConfig)
+	if err != nil {
+		return err
+	}
+
+	requiredPrivileges, err := computePrivileges(pd)
+	if err != nil {
+		return err
+	}
+
+	if !reflect.DeepEqual(privileges, requiredPrivileges) {
+		return errors.New("incorrect privileges")
 	}
 
 	pluginID := stringid.GenerateNonCryptoID()
 	pluginDir := filepath.Join(pm.libRoot, pluginID)
 	if err := os.MkdirAll(pluginDir, 0755); err != nil {
 		logrus.Debugf("error in MkdirAll: %v", err)
-		return nil, err
+		return err
 	}
 
-	priv, err := pm.pull(ref, metaHeader, authConfig, pluginID)
-	if err != nil {
+	removePluginDir := func() {
```

Comment:
Its more elegant if we change this to a defer that acts on a named return error. Then we can remove L182 and L189.
```
defer func() {
		if err != nil {
			os.RemoveAll(pluginDir)...
		}
	}()
```